lib/config/config.go

```diff
@@ -255,25 +258,46 @@ func convertV12V13(cfg *Configuration) {
 	// Not using the ignore cache is the new default. Disable it on existing
 	// configurations.
 	cfg.Options.CacheIgnoredFiles = false
+
+	// Migrate UPnP -> NAT options
 	cfg.Options.NATEnabled = cfg.Options.DeprecatedUPnPEnabled
+	cfg.Options.DeprecatedUPnPEnabled = false
 	cfg.Options.NATLeaseM = cfg.Options.DeprecatedUPnPLeaseM
+	cfg.Options.DeprecatedUPnPLeaseM = 0
 	cfg.Options.NATRenewalM = cfg.Options.DeprecatedUPnPRenewalM
+	cfg.Options.DeprecatedUPnPRenewalM = 0
 	cfg.Options.NATTimeoutS = cfg.Options.DeprecatedUPnPTimeoutS
-	if cfg.Options.DeprecatedRelaysEnabled {
-		cfg.Options.ListenAddresses = append(cfg.Options.ListenAddresses, cfg.Options.DeprecatedRelayServers...)
-		// Replace our two fairly long addresses with 'default' if both exist.
-		var newAddresses []string
-		for _, addr := range cfg.Options.ListenAddresses {
-			if addr != "tcp://0.0.0.0:22000" && addr != "dynamic+https://relays.syncthing.net/endpoint" {
-				newAddresses = append(newAddresses, addr)
+	cfg.Options.DeprecatedUPnPTimeoutS = 0
+
+	// Migrate listen & relay addresses.
+	if len(cfg.Options.ListenAddresses) == 1 && cfg.Options.ListenAddresses[0] == "tcp://0.0.0.0:22000" {
+		// This was the old default value. Set it to the new default value.
+		cfg.Options.ListenAddresses[0] = "default"
+
+		// If there were custom relays in the config before, add them now as
+		// well. We skip the default relay address as that is already
+		// covered by the "default" value.
+		for _, addr := range cfg.Options.DeprecatedRelayServers {
+			if addr != DefaultRelayListenAddresses[0] {
+				cfg.Options.ListenAddresses = append(cfg.Options.ListenAddresses, addr)
 			}
 		}
-
-		if len(newAddresses)+2 == len(cfg.Options.ListenAddresses) {
-			cfg.Options.ListenAddresses = append([]string{"default"}, newAddresses...)
+	} else {
+		// A custom listen address was set. We keep that around, and add in
+		// any relay servers that were configured (including the default
+		// pool). We *don't* do this if relays were disabled, though.
+		if cfg.Options.RelaysEnabled {
+			cfg.Options.ListenAddresses = append(cfg.Options.ListenAddresses, cfg.Options.DeprecatedRelayServers...)
 		}
 	}
-	cfg.Options.DeprecatedRelaysEnabled = false
+
```

Comment:
I think that we should still replace stuff with default if one of the old listen addresses was 22000?